### PR TITLE
Dashcontroller

### PR DESCRIPTION
**Pull request by tomcrane**
This is a huge PR but it has many dependencies to go from the demo dashboard page to a half-working dashboard.
Rather than have a huge comment I have written up the progress notes here:

https://github.com/wellcomecollection/iiif-builder/blob/dashcontroller/migration-notes/dashcontroller/README.md

### CODE REVIEW

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+using System.Threading.Tasks;
+using Wellcome.Dds.AssetDomain.Dlcs.Ingest;
+using Wellcome.Dds.AssetDomain.Mets;
+
+namespace Wellcome.Dds.AssetDomainRepositories.Ingest
+{
+
+    public class CloudServicesIngestRegistry : IIngestJobRegistry
+    {
+        private readonly IMetsRepository _metsRepository;
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:15]: underscore

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                .Take(number)
+                .Include(j => j.DlcsBatches)
+                .ToArray(); // The operation cannot be completed because the DbContext has been disposed. Want to defer this
+        }
+
+        public DlcsIngestJob GetJob(int id)
+        {
+            return ddsInstrumentationContext.DlcsIngestJobs
+                .Include(j => j.DlcsBatches)
+                .FirstOrDefault(j => j.Id == id);
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:39]: Would `.SingleOrDefault(j => j.Id == id);` be safer here? Assuming there shouldn't be >1 with same Id

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                .Where(j => j.StartProcessed == null);
+
+            if (after.HasValue)
+            {
+                return query.Where(j => j.Created > after.Value).ToArray();
+            }
+            return query.ToArray();
+        }
+
+        public IEnumerable<DlcsIngestJob> GetProblems(DateTime? after)
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:56]: Can `after` go? Not used here and commented out in the original version.

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            if (immediateStart)
+            {
+                job.StartProcessed = DateTime.Now;
+            }
+            return job;
+        }
+
+        // TODO: Making this async needs a code review (compare CloudServicesIngestRegistry and its interface with dds-ecosystem)
+        // The following few mthods all need reviewing
+        public async IAsyncEnumerable<DlcsIngestJob> RegisterImagesForImmediateStartAsync(string identifier)
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:106]: Can this become 

```cs
public IAsyncEnumerable<DlcsIngestJob> RegisterImagesForImmediateStartAsync(string identifier) 
  => RegisterImagesInternalAsync(identifier, false, true);
```

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+        public async Task<DlcsIngestJob[]> RegisterImagesAsync(string identifier, bool useInitialOrigin = false)
+        {
+            // Can this not return an array? Why does it need to?
+            var jobs = new List<DlcsIngestJob>();
+            await foreach (var job in RegisterImagesInternalAsync(identifier, useInitialOrigin, false))
+            {
+                jobs.Add(job);
+            }
+            return jobs.ToArray();
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:122]: This method can use `System.Linq.Async` nuget and then be simplified to `return await RegisterImagesInternalAsync(identifier, useInitialOrigin, false).ToArrayAsync();`. 

Unfortunately adding that nuget package causes mayhem when using alongside `DbSet` - not sure if it's worth changing? https://github.com/dotnet/efcore/issues/18124

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                    {
+                        existingQuery = existingQuery.Where(j => j.IssuePart == job.IssuePart);
+                    }
+                    foreach (var dlcsIngestJob in existingQuery)
+                    {
+                        ddsInstrumentationContext.DlcsIngestJobs.Remove(dlcsIngestJob);
+                    }
+                    // now add the new one
+                    ddsInstrumentationContext.DlcsIngestJobs.Add(job);
+                    ddsInstrumentationContext.SaveChanges();
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:168]: By default `.SaveChanges()` will run in a Transaction so shouldn't need to manually create, unless `IsolationLevel.Serializable` is needed? See: https://docs.microsoft.com/en-us/ef/core/saving/transactions#default-transaction-behavior

File: src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs [hunk 7]
```diff
@@ -0,0 +1,840 @@
...
+            IStatusProvider statusProvider,
+            Synchroniser synchroniser,
+            IDatedIdentifierProvider recentlyDigitisedIdentifierProvider,
+            IWorkStorageFactory workStorageFactory,
+            CacheBuster cacheBuster,
+            IOptions<DdsOptions> ddsOptions,
+            IOptions<DlcsOptions> dlcsOptions,
+            IDds dds,
+            IWebHostEnvironment webHostEnvironment,
+            IWorkflowCallRepository workflowCallRepository)
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs:64]: Not for now but just looking at dep list this is doing too much

File: src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+
+        public ActionResult Status(int page = 1)
+        {
+            var problemJobs = jobRegistry.GetProblems(statusProvider.EarliestJobToTake);
+            Page<ErrorByMetadata> errorsByMetadataPage;
+            try
+            {
+                errorsByMetadataPage = dashboardRepository.GetErrorsByMetadata(page);
+            }
+            catch (Exception)
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs:99]: Worth catching and logging exception?

File: src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs [hunk 9]
```diff
@@ -0,0 +1,840 @@
...
+            return View("ManifestationError");
+        }
+
+
+        private async Task<IDigitisedCollection> GetCachedCollectionAsync(string identifier)
+        {
+            var coll = await cache.GetCached(
+                CacheSeconds,
+                CacheKeyPrefix + identifier,
+                async () => (await dashboardRepository.GetDigitisedResourceAsync(identifier)));
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs:314]: `() => dashboardRepository.GetDigitisedResourceAsync(identifier));`? Not sure how `ISimpleCache` will handle a `Task<T>` as it's not awaiting it...?

File: src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs [hunk 10]
```diff
@@ -0,0 +1,840 @@
...
+            }
+            catch (Exception ex)
+            {
+                return RedirectToAction("ManifestationSearchError", q);
+            }
+        }
+
+        public async Task<ActionResult> Sync(string id)
+        {
+            return await CreateAndProcessJobsAsync(id, false, false, "Sync");
```
Comment (donaldgray) [inline src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs:419]: Same with next few calls, these pass throughs can omit `await` and `async`

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Dashboard/DashboardRepository.cs [hunk 11]
```diff
@@ -716,5 +716,18 @@ public IEnumerable<IngestAction> GetRecentActions(int count, string user = null)
         {
             return dlcs.GetDlcsQueueLevel();
         }
+
+        public AVDerivative[] GetAVDerivatives(IDigitisedManifestation digitisedManifestation)
+        {
+            var derivs = new List<AVDerivative>();
+            if (digitisedManifestation.MetsManifestation.Type == "Video" || digitisedManifestation.MetsManifestation.Type == "Video")
```
[the remark this answers is not in the document]
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Dashboard/DashboardRepository.cs:723]: yes

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs (hunk 1 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs]: fixed in that unpushed commit

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs (hunk 2 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:39]: yes, have changed

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs (hunk 3 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:56]: There is a use in another utility:
https://github.com/wellcomelibrary/dds-ecosystem/blob/new-storage-service/wellcome-dds/DlcsBackgroundReconciler/Program.cs#L164

However, I think we should still drop this and reintroduce if we need to go back.

Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:56]: However... I will turn the Take(..) arg into a parameter.

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs (hunk 4 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:107]: "Cannot return a value from an iterator. Use the yield return statement... " is the error I get when I try to do that.

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs (hunk 5 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:123]: I think stay safe.. I don't mind the verbosity there.

File: src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs (hunk 7 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs:64]: Yes agreed, this Controller is doing all the work and needs to be split up. Which is not quite the same problem but they are related.

File: src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs (hunk 6 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.AssetDomainRepositories/Ingest/CloudServicesIngestRegistry.cs:170]: I'll leave this for now but add a comment to revisit.
I can't remember why this is in a transaction.

File: src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs (hunk 8 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs:99]: Oddly considering the dependency list comment above - this controller doesn't yet have a logger!
Going to leave for now

File: src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs (hunk 9 above)
Comment (tomcrane) [reply src/Wellcome.Dds/Wellcome.Dds.Dashboard/Controllers/DashController.cs:314]: Added some comments here for later attention

**Pull request merged**